pkg/applicationserver/applicationserver.go

```diff
@@ -588,6 +592,202 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
+func (as *ApplicationServer) replaceDownlinkQueue(ctx context.Context, dev *ttnpb.EndDevice, link *link, previousSession *ttnpb.Session, downlinks []*ttnpb.ApplicationDownlink, nextAFCntDown uint32, replacePending bool) error {
+	return as.changeDownlinkQueue(ctx, dev, link, func(ctx context.Context, dev *ttnpb.EndDevice) (newCurrentQueue []*ttnpb.ApplicationDownlink, newPendingQueue []*ttnpb.ApplicationDownlink, err error) {
+		downlinks, unmatched := ttnpb.PartitionDownlinksBySessionKeyIDEquality(previousSession.SessionKeyID, downlinks...)
+		for _, item := range unmatched {
+			log.FromContext(ctx).WithFields(log.Fields(
+				"f_port", item.FPort,
+				"f_cnt", item.FCnt,
+				"session_key_id", item.SessionKeyID,
+			)).Warn("Drop downlink message; unknown session key ID")
+			registerDropDownlink(ctx, dev.EndDeviceIdentifiers, item, errUnknownSession)
+		}
+		if replacePending && dev.PendingSession != nil {
+			newPendingQueue, err = as.migrateDownlinkQueue(ctx, dev.EndDeviceIdentifiers, downlinks, previousSession, dev.PendingSession, 1)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+		newCurrentQueue, err = as.migrateDownlinkQueue(ctx, dev.EndDeviceIdentifiers, downlinks, previousSession, dev.Session, nextAFCntDown)
+		if err != nil {
+			return nil, nil, err
+		}
+		return newCurrentQueue, newPendingQueue, nil
+	})
+}
+
+// migrateDownlinkQueue constructs a new downlink queue by decrypting the elements of the old queue and encrypting them using the new session.
+// If re-encrypting a message fails, the message is skipped.
+// This method mutates the new session's LastAFCntDown if any downlinks have been migrated.
+// This method does not change the contents of the old downlink queue.
+func (as *ApplicationServer) migrateDownlinkQueue(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, oldQueue []*ttnpb.ApplicationDownlink, oldSession *ttnpb.Session, newSession *ttnpb.Session, nextAFCntDown uint32) ([]*ttnpb.ApplicationDownlink, error) {
+	if oldSession.AppSKey == nil || newSession.AppSKey == nil {
+		return nil, errNoAppSKey
+	}
+	// TODO: Cache unwrapped keys (https://github.com/TheThingsNetwork/lorawan-stack/issues/36)
+	oldAppSKey, err := cryptoutil.UnwrapAES128Key(ctx, *oldSession.AppSKey, as.KeyVault)
+	if err != nil {
+		return nil, err
+	}
+	newAppSKey, err := cryptoutil.UnwrapAES128Key(ctx, *newSession.AppSKey, as.KeyVault)
+	if err != nil {
+		return nil, err
+	}
+	newSession.LastAFCntDown = nextAFCntDown - 1
+	logger := log.FromContext(ctx)
+	newQueue := make([]*ttnpb.ApplicationDownlink, 0, len(oldQueue))
+	for _, oldItem := range oldQueue {
+		logger := logger.WithFields(log.Fields(
+			"f_port", oldItem.FPort,
+			"f_cnt", oldItem.FCnt,
+			"session_key_id", oldItem.SessionKeyID,
+		))
+		frmPayload, err := crypto.DecryptDownlink(oldAppSKey, oldSession.DevAddr, oldItem.FCnt, oldItem.FRMPayload)
+		if err != nil {
+			logger.WithError(err).Warn("Drop downlink message; failed to decrypt")
```

Comment:
Use either `<what happened>;<action>` form or `<action>;<what happened>`, but don't mix both in different logging statements